scarb/src/core/manifest/toml_manifest.rs

```diff
@@ -150,6 +158,43 @@ pub struct TomlCairo {
     pub sierra_replace_ids: Option<bool>,
 }
 
+#[derive(Clone, Debug, Ord, PartialOrd, Eq, PartialEq, Hash, Serialize, Deserialize)]
+#[serde(into = "SmolStr", try_from = "SmolStr")]
+pub struct TomlProfileName(SmolStr);
+
+impl TomlProfileName {
+    pub fn try_new(name: SmolStr) -> Result<Self> {
+        // Some validations are required to ensure profile name can be used as a valid subdirectory.
+        ensure!(
+            name.as_str() != "",
+            "cannot use empty string as profile name"
+        );
+        ensure!(
+            !vec!["_", "package", "build", "debug", "doc", "test"].contains(&name.as_str()),
+            format!("profile name `{name}` is not allowed")
+        );
+        ensure!(
+            !name.to_string().starts_with(".."),
+            format!("profile name cannot start with `..` prefix")
+        );
+        let sanitized = sanitize(name.as_str());
+        ensure!(
+            sanitized == name.as_str(),
+            format!("profile name `{name}` is not allowed, consider using `{sanitized}` instead")
+        );
+        Ok(Self(name))
```

Comment:
1. Profile construction itself isn't guarding this. I think this should be part of `ProfileType` constructor. Target is validated here only because otherwise it would be possible to override `[lib]` with `[[targets.lib]]`
2. I think we are not strict enough here. Shouldn't profiles be always lowercase ascii alphanumeric, with `-` symbol only accepted?